Confirm README has all example config keys in it

aquarat opened the pull request:
Resolves:
> Parse the README file and expect all configuration keys of the `airseeker.example.json` to be present.

Closes #42

Siegrift commented on test/readme.test.ts:
```diff
@@ -0,0 +1,40 @@
+    const readmeData = readFileSync(path.join(__dirname, '../README.md')).toString();
+    expect(readmeData).toBeDefined();
+
+    // console.log(config);
```
remove

Siegrift commented on src/config/index.ts:
```diff
@@ -1,20 +1,19 @@
+    const rawConfig = JSON.parse(fs.readFileSync(join(configPath, airseekerConfigPath), 'utf8'));
     const secrets = parseSecrets(rawSecrets);
-    return configSchema.parseAsync(interpolateSecrets(rawConfig, secrets));
+    return configSchema.safeParse(interpolateSecrets(rawConfig, secrets));
```
Good catch. The config does not need to be parsed asynchronously.

However, you should use the `.parse` instead of `.safeParse` here (the response type is wrong now).

aquarat replied:
Weirdly it didn't fail, I've swapped it out 👌

Siegrift commented on test/readme.test.ts:
```diff
@@ -0,0 +1,40 @@
+    const keys = extractKeys(config).filter((item) => !z.coerce.number().safeParse(item).success);
+
+    const missingKeys = keys.filter((key) => readmeData.includes(`\n#### ${key}`));
+
```
This does exact opposite. We want to filter all keys that are not in README data. The interpolation check like this will not work. I would imagine somtehing like this:

```ts
    const readmeHashtagTitles = readmeData
      .split('\n')
      .map((line) => /^#+ `(?<title>.+)`/.exec(line)?.groups?.title)
      .filter((title) => !!title);
    const readmeBacktickTitles = readmeData
      .split('\n')
      .map((line) => /^`(?<title>[^`]+)`/.exec(line)?.groups?.title)
      .filter((title) => !!title);
    const readmeTitles = [...readmeHashtagTitles, ...readmeBacktickTitles];
```

to capture all names starting with hastags or the ones that start just with "\`" (such as "url").

and get missing ones like this:
```ts
    const missingKeys = configKeys.filter(
      (configKey) => !readmeTitles.some((readmeTitle) => readmeTitle!.includes(configKey))
    );
```

but feel free to come up with a nicer heuristic.

aquarat replied:
Sorry about that, I didn't test the inverse. I've added in the negation and changed the README a bit. If you prefer I can go with your approach above instead 👌

Siegrift commented on test/readme.test.ts:
```diff
@@ -0,0 +1,40 @@
+    expect(readmeData).toBeDefined();
+
+    // console.log(config);
+    const keys = extractKeys(config).filter((item) => !z.coerce.number().safeParse(item).success);
```
You will need some manual filtering for object keys (e.g. hardhat) so might as well just create a the list of properties to filter out (I think only "31337" and "hardhat").

Siegrift commented on README.md:
```diff
@@ -126,75 +126,75 @@ The address of the Api3ServerV1 contract. If not specified, the address is loade
 A single element array for a beacon data. If the data feed is a beacon set, the array contains the data for all the
 beacons in the beacon set (in correct order).
 
-`dataFeedIdToBeacons<DATA_FEED_ID>[n]`
```
I ran out of nesting options so I just put it in backticks. I think what you did (make it title with #####) is better so I'd keep it.

Siegrift commented on README.md:
````diff
@@ -48,30 +48,30 @@ LOG_FORMAT=json
 ```
 
-#### `LOGGER_ENABLED`
+### `LOGGER_ENABLED`
````
These titles were nested because the `### Environment variables` already uses `###`.

aquarat replied:
No problem, I've reverted most of the README changes.

Siegrift commented on src/config/index.ts:
```diff
@@ -1,20 +1,23 @@
-  const configPath = join(__dirname, '../../config');
-  const rawSecrets = dotenv.parse(readFileSync(join(configPath, 'secrets.env'), 'utf8'));
+export const loadConfig = (
+  configPath = join(__dirname, '../../config'),
```
Do we need these parameters? I think the implementation works even when the values are hadcoded.

aquarat commented:
I've reverted the README changes, implemented your regex based search and mocked the readFileSync calls so the config loader will work with the example config without require extra arguments :) @Siegrift

Siegrift reviewed (approved):
👍 LGTM
